platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsToolingApiGradleLifecycleIntegrationTest.groovy

```diff
@@ -0,0 +1,134 @@
+        given:
+        withSomeToolingModelBuilderPluginInBuildSrc()
+        settingsFile << """
+            include("a")
+
+            gradle.lifecycle.beforeProject {
+                println("Callback before \$it")
+            }
+        """
+        buildFile << """
+            plugins.apply(my.MyPlugin)
+        """
+        file("a/build.gradle") << """
+            plugins.apply(my.MyPlugin)
+        """
+
+        when:
+        withIsolatedProjects()
+        def model = runBuildAction(new FetchCustomModelForEachProject())
+
+        then:
+        model.size() == 2
+        model[0].message == "It works from project :"
+        model[1].message == "It works from project :a"
+
+        fixture.assertStateStored {
+            projectConfigured(":buildSrc")
+            buildModelCreated()
+            modelsCreated(":", ":a")
+        }
+
+        outputContains("Callback before root project 'root'")
+        outputContains("Callback before project ':a'")
+
+
+        when:
+        withIsolatedProjects()
+        def model2 = runBuildAction(new FetchCustomModelForEachProject())
+
+        then:
+        model2.size() == 2
+        fixture.assertStateLoaded()
+        outputDoesNotContain("Callback before")
+
+
+        when:
+        buildFile << """
+            myExtension.message = "updated message for root"
+        """
+
+        withIsolatedProjects()
+        def model3 = runBuildAction(new FetchCustomModelForEachProject())
+
+        then:
+        model3.size() == 2
+        model3[0].message == "updated message for root"
+        model3[1].message == "It works from project :a"
+
+        fixture.assertStateUpdated {
+            fileChanged("build.gradle")
+            projectConfigured(":buildSrc")
+            modelsCreated(":")
+            modelsReused(":a", ":buildSrc")
+        }
+
+        outputContains("Callback before root project 'root'")
+        outputDoesNotContain("Callback before project ':a'")
+
+
+        when:
+        withIsolatedProjects()
+        def model4 = runBuildAction(new FetchCustomModelForEachProject())
+
+        then:
+        model4.size() == 2
+        fixture.assertStateLoaded()
+        outputDoesNotContain("Callback before")
+
+
+        when:
+        file("a/build.gradle") << """
+            myExtension.message = "updated message for :a"
+        """
+
+        withIsolatedProjects()
+        def model5 = runBuildAction(new FetchCustomModelForEachProject())
+
+        then:
+        model5.size() == 2
+        model5[0].message == "updated message for root"
+        model5[1].message == "updated message for :a"
+
+        and:
+        fixture.assertStateUpdated {
+            fileChanged("a/build.gradle")
+            projectConfigured(":buildSrc")
+            projectConfigured(":")
+            modelsCreated(":a")
+            modelsReused(":", ":buildSrc")
+        }
+
+        outputContains("Callback before root project 'root'")
```

Comment:
🤔 I wasn't expecting the root project to be reconfigured due to a change to a subproject, interesting...

Reply:
I discussed this with Adam. The current behavior is to reconfigure the projects in the project path up to the root because of the `pushing down of configuration` that's common such as _inherited properties_ (to be replaced by _Shared Data_ eventually), _subprojects_ configuration blocks (which might still make sense for the incremental adoption story), etc.